packages/calypso-e2e/src/lib/flows/difm-flow.ts

```diff
@@ -0,0 +1,115 @@
+import { Page } from 'playwright';
+import { DataHelper } from '../..';
+
+const selectors = {
+	// General
+	button: ( text: string ) => `button:text("${ text }")`,
+
+	// Start page
+	startContainer: '.signup.is-do-it-for-me',
+
+	// Options page
+	siteNameInput: '#siteTitle',
+	taglineInput: '#tagline',
+
+	// Social page
+	facebookInput: 'input[name=facebookUrl]',
+
+	// Design page
+	designPickerContainer: '.design-picker',
+
+	//Page picker page
+	pagePickerContainer: '.signup__step .is-difm-page-picker',
+
+	// Checkout page
+	checkoutContainer: '.checkout__content',
+};
+
+/**
+ * Class encapsulating the flow when starting a new do-it-for-me site order.
+ */
+export class DIFMFlow {
+	/**
+	 * Constructs an instance of the flow.
+	 *
+	 * @param {Page} page The underlying page.
+	 */
+	constructor( private page: Page ) {}
+
+	/**
+	 * Given text, click on the button's first instance with the text.
+	 *
+	 * @param {string} text User-visible text on the button.
+	 */
+	async clickButton( text: string ): Promise< void > {
+		const selector = selectors.button( text );
+		await this.page.click( selector );
+	}
+
+	/**
+	 * Go to first setup page.
+	 */
+	async startSetup(): Promise< void > {
+		await this.page.goto(
+			DataHelper.getCalypsoURL(
+				'/start/do-it-for-me/new-or-existing-site?flags=signup/redesigned-difm-flow'
+			)
+		);
+		await this.validateSetupPage();
+	}
+
+	/**
+	 * Validates that we've landed on the setup page.
+	 */
+	async validateSetupPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.startContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the options page.
+	 */
+	async validateOptionsPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.siteNameInput );
+	}
+
+	/**
+	 * Validates that we've landed on the social page.
+	 */
+	async validateSocialPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.facebookInput );
+	}
+
+	/**
+	 * Validates that we've landed on the design picker page.
+	 */
+	async validateDesignPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.designPickerContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the page picker page.
+	 */
+	async validatePagePickerPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.pagePickerContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the checkout page.
+	 */
+	async validateCheckoutPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.checkoutContainer, { timeout: 60000 } );
+	}
```

Comment:
Given that we seem to be validating a page load over 6 different pages that each only differ by a selector that is passed in, what might be an approach to reduce the amount of repetitive code?

Reply:
Agree, but I do think that we should not check for header text directly inside the test. I extracted the main selector call to a reusable method and just have methods for every page containing a validation call with just the page header. This way, if we ever have to change the method call behind it, there is only one place to look at.

WDYT @worldomonation?

Reply:
> I do think that we should not check for header text directly inside the test.

I do agree - the test does not need to be the one performing the check.

My suggestion was towards a unified method, something like:

```typescript
async validatePage( name: DIFMPages ): {
    ....
}
```

Where the `DIFMPages` is a type/interface of the different DIFM pages. 

Typically, I avoid making a function per action, instead preferring to have a generic function that accepts a parameter, and switch the action based on the parameter value. 

An example of this is in the `sidebar-component.ts` where I opted to have a generic function `navigate` instead of `visitMedia`, `visitPages`, `visitPosts`, etc.

I recognize there is no right answer here and it comes down design philosophy and preference, but I wanted to know if you felt strongly either for or against that.

Reply:
To be honest, I do not have a strong opinion for or against it. I do like to have less clutter inside a class but I think that this makes it not really smaller. I am concerned that it makes the class more complicated than it needs to be in my case. It would be much easier for someone that has not yet taken a look at the automation framework to see how they can change something when looking at my current implementation. 🤔

Reply:
I like to obey by KISS (keep it simple stupid) in most cases, but as I said I do not have a fixed opinion on this 🤷🏻‍♂️